base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -2172,7 +2183,20 @@ fn insert_orphan_and_find_new_tips<T: BlockchainBackend>(
         let timestamp = EpochTime::from(h.timestamp());
         prev_timestamps.push(timestamp);
     }
-    validator.validate(db, &block.header, parent.header(), &prev_timestamps, None)?;
+    let result = validator.validate(db, &block.header, parent.header(), &prev_timestamps, None);
+    if let Err(e) = &result {
+        match e {
+            // future timelimit validation can succeed at a later time. As the block is not yet valid, we discard it
+            // for now and ban the peer, but wont blacklist the block.
+            BlockHeaderError(BlockHeaderValidationError::InvalidTimestampFutureTimeLimit) => {},
+            _ => {
+                let mut txn = DbTransaction::new();
+                txn.insert_bad_block(block.header.hash(), block.header.height);
+                db.write(txn)?;
+            },
+        }
+    }
+    result?;
     let achieved_target_diff = difficulty_calculator.check_achieved_and_target_difficulty(db, &block.header)?;
```

Comment:
nit: match on the result rather than `result?` which desugars into another match. It also makes it clearer IMO that we are still exiting the function early if FTL fails. 
Note that the validator returns the `achieved_target_diff` so it doesn't have to be done again below

```suggestion
    let achieved_target_diff = match result {
        Ok(achieved_diff) => achieved_diff,
        Err(e @ValidationError::BlockHeaderError(BlockHeaderValidationError::InvalidTimestampFutureTimeLimit)) => return Err(e.into()),
        Err(e @ValidationError::FatalStorageError(_)) => return Err(e.into()),
        Err(e) => {
            let mut txn = DbTransaction::new();
            txn.insert_bad_block(block.header.hash(), block.header.height);
            db.write(txn)?; 
            return Err(e.into());
        }
    };
```

Reply:
Interestingly, if I add this code below, we get a mismatch in achieved difficulty in `it_correctly_handles_duplicate_blocks` test
```rust
    let d2 = difficulty_calculator.check_achieved_and_target_difficulty(db, &block.header)?;
    assert_eq!(achieved_target_diff.achieved(), d2.achieved());
```
 EDIT: ah it's a symptom of us using a mock validator in tests or use the real validator 
We'll need to change the mock to return valid difficulty
```rust
impl<B: BlockchainBackend> HeaderChainLinkedValidator<B> for MockValidator {
    fn validate(
        &self,
        db: &B,
        header: &BlockHeader,
        _: &BlockHeader,
        _: &[EpochTime],
        _: Option<Difficulty>,
    ) -> Result<AchievedTargetDifficulty, ValidationError> {
        if self.is_valid.load(Ordering::SeqCst) {
            let consensus = create_consensus_rules();
            let difficulty_calculator = DifficultyCalculator::new(consensus, Default::default());
            difficulty_calculator.check_achieved_and_target_difficulty(db, header)
        } else {
            Err(ValidationError::custom_error(
                "This mock validator always returns an error",
            ))
        }
    }
}
```